From df5cde447874ab006a19bb19a49f22c747c25a46 Mon Sep 17 00:00:00 2001
From: Yao Zi <ziyao@disroot.org>
Date: Wed, 24 Sep 2025 05:05:35 +0000
Subject: [PATCH] slbt_ar_merge_archives(): Count nametable's size from
 filenames

When an archive contains two or more objects with the same filename,
it's possible for an ar implementation to reuse part of the nametable
for the latter objects, which is LLVM ar's behavior.

For slibtool, we don't have such deduplication mechanisms, instead, a
new entry is always created in nametable if the corresponding object
requires a long name. This breaks the assumption that the result
nametable's size is the sum of all input archives' nametable, if one of
the input archive contains several objects with the same name and part
of its nametable is reused, leading to malformed archive structure.

Let's calculate size of the result nametable as the sum of all long-name
archive members' names instead.

Signed-off-by: Yao Zi <ziyao@disroot.org>
---
 src/arbits/slbt_archive_merge.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/arbits/slbt_archive_merge.c b/src/arbits/slbt_archive_merge.c
index e62ba121c333..adca7138e548 100644
--- a/src/arbits/slbt_archive_merge.c
+++ b/src/arbits/slbt_archive_merge.c
@@ -264,8 +264,6 @@ int slbt_ar_merge_archives(
 					break;
 
 				case AR_MEMBER_ATTR_NAMESTRS:
-					snamestrs += meminfo->ar_object_size;
-
 					if (!arnames)
 						arnames = meminfo;
 
@@ -279,6 +277,12 @@ int slbt_ar_merge_archives(
 					smembers ^= 1;
 					nmembers++;
 
+					if ((meminfo->ar_file_header.ar_header_attr & AR_HEADER_ATTR_SYSV) &&
+					    (meminfo->ar_file_header.ar_header_attr & AR_HEADER_ATTR_NAME_REF)) {
+						snamestrs += strlen(meminfo->ar_file_header.ar_member_name);
+						snamestrs += 2; /* AR_OBJ_PADDING and '/' */
+					}
+
 					break;
 			}
 		}
-- 
2.51.0

